docs/config.rst
Configuration
=============


+------------------+------------------------+-----------------------------------------------------------------+
| **Parameters**   | **Values**             | **Description**                                                 |
+------------------+------------------------+-----------------------------------------------------------------+
| audio.samplerate | (int) default=44100    | Samplerate of the project in Hertz                              |
+------------------+------------------------+-----------------------------------------------------------------+
| audio.mel_bands  | (int) default=128      | Number of mel bands for the filter                              |
+------------------+------------------------+-----------------------------------------------------------------+
| audio.fmax       | (int) default=22050    | Frequency max in Hertz for the mel filter decomposition         |
+------------------+------------------------+-----------------------------------------------------------------+
| path.tmp         | (str) default='/tmp'   | Path to the temporary download folder                           |
+------------------+------------------------+-----------------------------------------------------------------+
| path.data        | (str) default='~/'     | Path to the data folder                                         |
+------------------+------------------------+-----------------------------------------------------------------+
| path.url_list    | (str) no default value | Path to the file containing the URLs of the data to download    |
+------------------+------------------------+-----------------------------------------------------------------+


HISTORY.rst
=======
History
=======

0.3.1 (2017-08-22)
------------------

Added
^^^^^

* `Issue #8`_: config file feature
* `Issue #11`_: Samplerate setting up through CLI
* `Issue #10`_: Spectrogram calculation
* `Issue #24`_: Add a 'config' subcommand to the CLI
* `Issue #25`_: Read a value from the config file
* `Issue #21`_: Script to verify the samplerate sanity of the database
* `Issue #30`_: Function to transform milliseconds to samples

Changed
^^^^^^^

* `Issue #23`_: Update of the documentation for the installation process
* `Issue #26`_: Update of the CLI subcommand 'getdata' 
* `Issue #32`_: Exception handling in config reading

Deprecated
^^^^^^^^^^

Nothing.

Removed
^^^^^^^

Nothing.

Fixed
^^^^^

* `Issue #9`_: trailing slash on data path
* `Issue #13`_: Update of the config file parameters
* `Issue #15`_: Crash when setting up the samplerate
* `Issue #22`_: Update of the dependency list
* `Issue #25`_: Update of the config file
* Update of docstrings
* Update of the documentation

Security
^^^^^^^^

* Update of the Coveralls library from 1.1 to 1.2.0

0.2.3 (2017-08-07)
------------------

Added
^^^^^

* Travis config file
* pytest suite
* CLI tests
* `issue #5`_: Package coverage for the development setup
* `issue #6`_: adding of a `clear_zip()` to clean the archive files

Changed
^^^^^^^

* Python 3.3 testing removed

Deprecated
^^^^^^^^^^

* Nothing

Removed
^^^^^^^

* Nothing

Fixed
^^^^^

* `Issue #4`_: invalid functools dependency
* `Issue #7`_: update of docstring `unzip_data()`

Security
^^^^^^^^

* Nothing


0.2.2 (2017-07-31)
------------------

Added
^^^^^

* nothing

Changed
^^^^^^^

* Updated of the ChangeLog (HISTORY.rst)

Deprecated
^^^^^^^^^^

* nothing

Removed
^^^^^^^

* nothing

Fixed
^^^^^

* nothing

Security
^^^^^^^^

* nothing

0.2.1 (2017-07-31)
------------------

Added
^^^^^
* nothing

Changed
^^^^^^^

* nothing

Deprecated
^^^^^^^^^^

* nothing

Removed
^^^^^^^

* nothing

Fixed
^^^^^

* unzip_data() url list issue
* download of temporary files in the right directory  

Security
^^^^^^^^

* nothing

0.2.0 (2017-07-31)
------------------

Added
^^^^^

* Adding of a documentation (with docstrings)
* CLI command to download and unzip data automatically
* creation of a python package
* configuration of Tox
* download() method in data class

Changed
^^^^^^^

* Use of RST instead of markdown for all the documentation
* development packages are now in requirements_dev.txt

Deprecated
^^^^^^^^^^
* nothing

Removed
^^^^^^^

* nothing

Fixed
^^^^^

* source files satisfy PEP8
* bug fix on getdata cli

Security
^^^^^^^^

* Update of all packages to their latest versions

0.1.0 (2017-07-25)
------------------

* First release as a package.


.. _Issue #4: https://github.com/mattberjon/asc-cnn/issues/4
.. _Issue #7: https://github.com/mattberjon/asc-cnn/issues/7
.. _Issue #5: https://github.com/mattberjon/asc-cnn/issues/5
.. _Issue #6: https://github.com/mattberjon/asc-cnn/issues/6
.. _Issue #9: https://github.com/mattberjon/asc-cnn/issues/9
.. _Issue #8: https://github.com/mattberjon/asc-cnn/issues/8
.. _Issue #13: https://github.com/mattberjon/asc-cnn/issues/13
.. _Issue #15: https://github.com/mattberjon/asc-cnn/issues/15
.. _Issue #11: https://github.com/mattberjon/asc-cnn/issues/11
.. _Issue #10: https://github.com/mattberjon/asc-cnn/issues/10
.. _Issue #22: https://github.com/mattberjon/asc-cnn/issues/22
.. _Issue #23: https://github.com/mattberjon/asc-cnn/issues/23
.. _Issue #24: https://github.com/mattberjon/asc-cnn/issues/24
.. _Issue #25: https://github.com/mattberjon/asc-cnn/issues/25
.. _Issue #26: https://github.com/mattberjon/asc-cnn/issues/26
.. _Issue #21: https://github.com/mattberjon/asc-cnn/issues/21
.. _Issue #30: https://github.com/mattberjon/asc-cnn/issues/30
.. _Issue #32: https://github.com/mattberjon/asc-cnn/issues/32


asc/cli.py
# -*- coding: utf-8 -*-

"""Console script for asc."""

import click
import os
import sys
from shutil import copyfile

from . import __version__
from . import audio
from . import data
from . import utils
from config import App


config_path = App.config('CONFIG_PATH')
config_url_list_path = App.config('URL_PATH')
config_tmp_path = App.config('TMP_PATH')
config_data_path = App.config('DATA_PATH')


def version_msg():
    """ Returns the program version, location and python version.
    """

    python_version = sys.version[:3]
    message = 'asc %(version)s (Python {})'
    return message.format(python_version)


@click.group(context_settings=dict(help_option_names=['-h', '--help']))
@click.version_option(
        __version__,
        '-V',
        '--version',
        message=version_msg(),
        help='Output the version of the application')
@click.pass_context
def main(ctx):
    """Console script for Acoustic Scene Classification (ASC).

    Need to set a better docstring.
    """
    pass


@main.command()
@click.option(
        '-f',
        '--filename')
@click.option(
        '-fs',
        '--frame-size',
        type=int,
        default=9,
        help='Frame size in samples. 1 frame=512 samples [default=9]')
@click.option(
        '-m',
        '--mel-bands',
        type=int,
        default=128,
        help='Number of mel bands to compute the dynamic spectrogram.\
                [default=128]')
@click.option(
        '-fm',
        '--frequency-max',
        type=int,
        default=22050,
        help='Frequency max to apply to the mel band in Hertz.\
                [default=22050]')
@click.option(
        '-D',
        '--debug',
        is_flag=True,
        help='Display each spectrograms per block')
def processing(frame_size, filename, mel_bands, frequency_max, debug):
    """ Set up the audio processing chain.
    """
    audio_path = utils.read_config('path', 'audio')
    for audio_item in os.listdir(audio_path):
        complete_path = os.path.join(audio_path, audio_item)
        audio.process_audio(
                complete_path,
                frame_size,
                mel_bands,
                frequency_max,
                display=debug)


@main.command()
@click.argument(
        'parameter',
        type=click.STRING)
@click.argument('value')
def config(parameter, value):
    """ Configure the project.

    Save the configuration into the configuration file [default=~/.ascrc]
    See the documentation for the available list of parameters and values.
    """
    # Separation the section from the option
    section, option = utils.conf_param_extract(parameter)
    utils.write_config(
            section,
            option,
            value)


@main.command()
def getdata():
    """Download dataset from the server.

    Download the dataset from the server and unzip the all in the default
    folder.
    """
    # Check the data folder. If the folder doesn't exist, we create it,
    # if it exist, then ask the user if we can override it before proceed
    # to do anything else and store the path in a config file.
    try:
        url_list = utils.read_config('path', 'url_list')
    except ValueError:
        url_list = config_url_list_path

    try:
        tmp_path = utils.read_config('path', 'tmp')
    except ValueError:
        tmp_path = config_tmp_path

    try:
        audio_path = utils.read_config('path', 'audio')
    except ValueError:
        audio_path = config_data_path + '/audio'

    # start download
    get_data = data.Data()
    file_list = get_data.file_to_list(url_list)
    get_data.download(file_list, tmp_path)

    # unzip the files
    get_data.unzip_data(file_list, tmp_path, tmp_path)

    #  move the files into the audio folder
    get_data.move_files(file_list, tmp_path, audio_path)


@main.command()
def setup():
    """ Create the necessary folders and update the config file

    This command asks for the root folder, create the necessary folders and
    update the configuration file accordingly.
    """
    var_name = 'Root path of the project'
    default_value = '~/asc-data'
    ret_val = utils.read_user_input(var_name, default_value)
    root_path = os.path.abspath(os.path.expanduser(ret_val))

    # Setup the paths
    archive_path = root_path + '/archives'
    audio_path = root_path + '/audio'
    specs_path = root_path + '/spectrograms'

    # Create the directories
    if not os.path.isdir(archive_path):
        os.makedirs(archive_path)
    if not os.path.isdir(audio_path):
        os.makedirs(audio_path)
    if not os.path.isdir(specs_path):
        os.makedirs(specs_path)

    # save the information in the config file
    section = 'path'
    utils.write_config(section, 'root', root_path)
    utils.write_config(section, 'archive', archive_path)
    utils.write_config(section, 'audio', audio_path)
    utils.write_config(section, 'spectrograms', specs_path)

    # Copy the URL file into the data folder
    url_list = root_path + '/url_list.txt'
    copyfile(config_url_list_path, url_list)
    utils.write_config(section, 'url_list', url_list)


if __name__ == "__main__":
    main()
